04 Strategy Library/Fama French Five Factors/03 Algorithm.html

```diff
@@ -0,0 +1,69 @@
+</p>
+<div class="section-example-container">
+<pre class="python">
+def FineSelectionFunction(self, fine):
+    '''Select securities with highest score on Fama French 5 factors'''
+
+    # Select stocks with these 5 factors:
+    # MKT -- Book value per share: Value
+    # SMB -- TotalEquity: Size
+    # HML -- Operation profit margin: Quality
+    # RMW -- ROE: Profitability
+    # CMA -- TotalAssetsGrowth: Investment Pattern
+    filtered = [x for x in fine if x.ValuationRatios.BookValuePerShare
+                                and x.FinancialStatements.BalanceSheet.TotalEquity
+                                and x.OperationRatios.OperationMargin.Value
+                                and x.OperationRatios.ROE
+                                and x.OperationRatios.TotalAssetsGrowth]
+
+    # Sort by factors
+    sortedByMkt = sorted(filtered, key=lambda x: x.ValuationRatios.BookValuePerShare, reverse=True)
+    sortedBySmb = sorted(filtered, key=lambda x: x.FinancialStatements.BalanceSheet.TotalEquity.Value, reverse=True)
+    sortedByHml = sorted(filtered, key=lambda x: x.OperationRatios.OperationMargin.Value, reverse=True)
+    sortedByRmw = sorted(filtered, key=lambda x: x.OperationRatios.ROE.Value, reverse=True)
+    sortedByCma = sorted(filtered, key=lambda x: x.OperationRatios.TotalAssetsGrowth.Value, reverse=False)
+
+    stockBySymbol = {}
+
+    # Get the rank based on 5 factors for every stock
+    for index, stock in enumerate(sortedByMkt):
+        mktRank = self.beta_m * index
+        smbRank = self.beta_s * sortedBySmb.index(stock)
+        hmlRank = self.beta_h * sortedByHml.index(stock)
+        rmwRank = self.beta_r * sortedByRmw.index(stock)
+        cmaRank = self.beta_c * sortedByCma.index(stock)
+        avgRank = np.mean([mktRank,smbRank,hmlRank,rmwRank,cmaRank])
+        stockBySymbol[stock.Symbol] = avgRank
+
+    sorted_dict = sorted(stockBySymbol.items(), key = lambda x: x[1], reverse = True)
+    symbols = [x[0] for x in sorted_dict]
+
+    # Pick the stocks with the highest scores to long
+    self.longSymbols= symbols[:self.num_long]
+    # Pick the stocks with the lowest scores to short
+    self.shortSymbols = symbols[-self.num_short:]
+
+    return self.longSymbols + self.shortSymbols   
+</pre>
+</div>
+
+<p>
+    In this example, the portfolio is rebalanced every 30 days and the backtest period runs from Jan 2010 to Aug 2019. You can improve upon this strategy by changing the fundamental factors, the weight of each factor and the rebalance frequency.
+</p>
+
+<div class="qc-embed-frame" style="display: inline-block; position: relative; width: 100%; min-height: 100px; min-width: 300px;">
```

Comment:
Create a new section called Results. 

Follow the pattern of the other strategy library tutorials

Reply:
Where is your method?